pkg/reconciler/broker/targets.go

```diff
@@ -0,0 +1,205 @@
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/encoding/prototext"
+	"google.golang.org/protobuf/proto"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/typed/core/v1"
+	"knative.dev/eventing/pkg/logging"
+)
+
+const (
+	targetsCMKey          = "targets"
+	targetsFileName       = "targets.txt"
+	targetsCMResyncPeriod = 10 * time.Second
+)
+
+type BrokerConfigUpdater interface {
+	UpdateBrokerConfig(context.Context, *config.Broker) error
+}
+
+type updateReq struct {
+	broker  *config.Broker
+	replyCh chan error
+}
+
+type targetsReconciler struct {
+	targetsName, targetsNS string
+	targets                *config.TargetsConfig
+	updateCh               chan updateReq
+	configMaps             v1.ConfigMapInterface
+	logger                 *zap.Logger
+}
+
+func NewTargetsReconciler(ctx context.Context, client kubernetes.Interface, targetsNS, targetsName string) BrokerConfigUpdater {
+	r := &targetsReconciler{
+		targetsName: targetsName,
+		targetsNS:   targetsNS,
+		updateCh:    make(chan updateReq),
+		configMaps:  client.CoreV1().ConfigMaps(targetsNS),
+		logger:      logging.FromContext(ctx),
+	}
+	if err := r.loadTargetsConfig(ctx); err != nil {
+		r.logger.Error("error loading targets config", zap.Error(err))
+		r.targets = new(config.TargetsConfig)
+	}
+	go r.reconcile(ctx)
+	return r
+}
+
+func (r *targetsReconciler) UpdateBrokerConfig(ctx context.Context, b *config.Broker) error {
+	if b.Name == "" || b.Namespace == "" {
+		return fmt.Errorf("broker missing namespace or name: %s", prototext.MarshalOptions{}.Format(b))
+	}
+
+	req := updateReq{
+		broker:  b,
+		replyCh: make(chan error, 1),
+	}
+	select {
+	case r.updateCh <- req:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	select {
+	case err := <-req.replyCh:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// loadTargetsConfig retrieves the targets ConfigMap and populates r.targets.
+func (r *targetsReconciler) loadTargetsConfig(ctx context.Context) error {
+	logger := r.logger.With(zap.String("targetsName", r.targetsName), zap.String("targetsNS", r.targetsNS))
+	logger.Debug("Loading targets config from configmap")
+	targetsConfigMap, err := r.configMaps.Get(r.targetsName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.targets = &config.TargetsConfig{
+				Brokers: make(map[string]*config.Broker),
+			}
+			return nil
+		}
+		return fmt.Errorf("error getting targets ConfigMap: %w", err)
+	}
+
+	targets := new(config.TargetsConfig)
+	if err := proto.Unmarshal(targetsConfigMap.BinaryData[targetsCMKey], targets); err != nil {
+		return err
+	}
+	r.logger.Debug("Loaded targets config from ConfigMap",
+		zap.String("resourceVersion", targetsConfigMap.ResourceVersion))
+
+	r.targets = targets
+	return nil
+}
+
+func (r *targetsReconciler) reconcile(ctx context.Context) {
```

Comment:
This approach uses a single goroutine to guard access to state. This achieves more or less the same thing as a mutex but has the advantage that code executed within this goroutine does not need to worry about locking. Since all communication happens over channels it easy for the main goroutine to select on a resync channel for example without worrying about locking. Channels also allow for cancellation where mutexes do not.

The end result is that this reconciler combines the communication of state changes and serialization of config updates that was provided by the old targets updater goroutine with the protection of mutable state provided by in memory targets data structure.

Reply:
I'm likely biased. But I really don't see a point of forcing channel adoption in this case. This approach here seems to leak a lot of internals of targets config while IMO the reconciler should only worry about reading/writing the configmap data. It would be much easier for me to agree with such a change if update logic can be better wrapper within the targets config lib.